datafusion/optimizer/src/filter_push_down.rs

```diff
@@ -248,6 +249,128 @@ fn get_pushable_join_predicates<'a>(
         .unzip()
 }
 
+// examine OR clause to see if any useful clauses can be extracted and push down.
```

Comment:
I don't think this transformation is correct. In particular, I don't think the results will always be the same

Schematically, we have this type of predicate (that is being evaluated during the join)

```
(A AND B) OR (C AND D)
```

This transformation proposes adding another `(A OR B)` clause (evaluated before the join), so effectively

```
 ((A AND B) OR (C AND D)) AND (A OR B)
```

In order to do this transformation, the boolean statements must be equivalent for all inputs. 

However, a counter example is
```
A: false, B: false, C: true, D: true
```

In this case, the original predicate would be true, but the rewrite would be false

Here is the program I wrote to generate the entire truth table: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=334938478775ba3cd55e7c400ea89b06

Reply:
This transformation should extract at least one quals from each sub-clauses of OR, else do nothing.

(A AND B) OR (C AND D)

will be transformed to

((A AND B) OR (C AND D)) AND (A OR C)   

OR

((A AND B) OR (C AND D)) AND ((A AND B) OR C)   

OR 

do nothing.

Reply:
I see -- thanks -- I checked those rewrites and https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3b41b0409c8ecf4df0027f323668e0db they do look good to me